src/navigators/createMaterialTopTabNavigator.js

```diff
@@ -113,18 +133,109 @@ class TabView extends React.PureComponent<Props> {
 
   _renderPanPager = props => <TabViewPagerPan {...props} />;
 
-  _renderScene = ({ route, focused }) => {
-    const { renderScene, animationEnabled, swipeEnabled } = this.props;
+  componentWillMount() {
+    const { animationEnabled } = this.props;
 
-    if (animationEnabled === false && swipeEnabled === false) {
-      return (
-        <ResourceSavingScene isFocused={focused}>
-          {renderScene({ route })}
-        </ResourceSavingScene>
+    // We rely on action listener to animate from the current index to the next one, no need to listen without animation
+    if (animationEnabled) {
+      this._actionListener = this.props.navigation.addListener(
+        'action',
+        this._onAction
       );
     }
+  }
+
+  componentWillUnmount() {
+    if (this._actionListener) {
+      this._actionListener.remove();
+    }
+
+    if (this._transitionTimeout) {
+      clearTimeout(this._transitionTimeout);
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (
+      nextProps.navigation.state.index !== this.props.navigation.state.index
+    ) {
+      const { index } = nextProps.navigation.state;
+
+      this.setState(state => ({
+        loaded: state.loaded.includes(index)
+          ? state.loaded
+          : [...state.loaded, index],
+      }));
+    }
+  }
+
+  _onAction = payload => {
+    if (payload.action.type === NavigationActions.NAVIGATE) {
+      this.setState({
+        transitioningFromIndex:
+          (payload.lastState && payload.lastState.index) || 0,
+      });
+    } else if (payload.action.type === 'Navigation/COMPLETE_TRANSITION') {
```

Comment:
AFAIK tab navigator doesn't have fire transition actions. What does this do?

Reply:
It was before, but you're right current version (2.0.1) doesn't. I'll look at `onAnimationEnd` prop that's exactly what I need.